### PR TITLE
Best Practices-adityapurohit

### PR DESCRIPTION
**Pull request by adityapurohit1771**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+cjapp/.env
+cjapp/__pycache__/
+restapi/__pycache__/
```
Comment (kushal-ti) [inline .gitignore:3]: you can just use **/__pycache__/ to automatically add all pycache folders to .gitingore

File: restapi/views.py [hunk 2]
```diff
@@ -1,64 +1,72 @@
...
-from restapi.models import *
-from restapi.serializers import *
-from restapi.custom_exception import *
-
+from restapi.models import Category, Groups, Expenses, UserExpense
+from restapi.serializers import UserSerializer, CategorySerializer, GroupSerializer, ExpensesSerializer
+from restapi.custom_exception import UnauthorizedUserException
 
+#Constants
+EXECUTOR_TIMEOUT = 60
```
Comment (kushal-ti) [inline restapi/views.py:26]: should go directly into the constants file instea

File: restapi/views.py [hunk 3]
```diff
@@ -1,64 +1,72 @@
...
 
 def index(_request):
     return HttpResponse("Hello, world. You're at Rest.")
 
 
 @api_view(['POST'])
 def logout(request):
+    logging.info(f"Deleting auth_token for {request.user.id}")
```
Comment (kushal-ti) [inline restapi/views.py:34]: I'm not completely sure about this, but most likely, this won't log anything. python logging doesn't work like this, you need to call logging.basicConfig() at some point to configure a console logger. Also, it's preferable to use the getLogger method to get a separate logger instance, with custom attributes for logger name and default level than just using logging

File: restapi/views.py [hunk 4]
```diff
@@ -1,64 +1,72 @@
...
     request.user.auth_token.delete()
+    logging.info(f"auth_token deleted. for {request.user.id}")
     return Response(status=status.HTTP_204_NO_CONTENT)
 
 
 @api_view(['GET'])
 def balance(request):
+    logging.info("Getting Balances")
```
Comment (kushal-ti) [inline restapi/views.py:42]: add docstrings across the codebase

File: restapi/views.py [hunk 5]
```diff
@@ -180,24 +207,37 @@ def get_queryset(self):
...
-    num_threads = data['parallelFileProcessingCount']
-    log_files = data['logFiles']
+def log_processor(request):
+    logging.info("proccesing log data")
+    try:
+        data = request.data
+        num_threads = data['parallelFileProcessingCount']
+        log_files = data['logFiles']
+    except:
+        logging.error("Failed to fetch data for log proccesing")
```
Comment (kushal-ti) [inline restapi/views.py:217]: literally the first example i've seen so far where someone did exception handling properly, kudos

File: restapi/views.py [hunk 6]
```diff
@@ -227,46 +267,48 @@ def aggregate(cleaned_logs):
...
+def reader(url:str, timeout:int)->str:
     with urllib.request.urlopen(url, timeout=timeout) as conn:
-        return conn.read()
+        data:str = conn.read()
+        data = data.decode('utf-8')
+        return data
 
 
-def multiThreadedReader(urls, num_threads):
+def multi_thread_reader(urls:list, num_threads:int)->list:
```
Comment (kushal-ti) [inline restapi/views.py:304]: would be better to clump all log_processor related functions into a separate utility class than having them all in views.py

File: restapi/views.py [hunk 7]
```diff
@@ -227,46 +267,48 @@ def aggregate(cleaned_logs):
...
+                key = f"{hours}:45-{hours+1}:00"
         elif minutes >= 30:
-            key = "{:02d}:30-{:02d}:45".format(hours, hours)
+            key = f"{hours}:30-{hours}:45"
         elif minutes >= 15:
-            key = "{:02d}:15-{:02d}:30".format(hours, hours)
+            key = f"{hours}:15-{hours}:30"
         else:
-            key = "{:02d}:00-{:02d}:15".format(hours, hours)
+            key = f"{hours}:00-{hours}:15"
```
Comment (kushal-ti) [inline restapi/views.py:289]: extra points for using f-strings :+1:

File: restapi/serializers.py [hunk 8]
```diff
@@ -33,7 +33,7 @@ class Meta(object):
 
 
 class UserExpenseSerializer(ModelSerializer):
-    class Meta(object):
+    class Meta():
         model = UserExpense
         fields = ['user', 'amount_owed', 'amount_lent']
```
Comment (kushal-ti) [inline restapi/serializers.py:39]: remove dead code from expense serializer

**Review by kushal-ti — CHANGES_REQUESTED** [review]
Good attempt. Make some of the changes i've requested and i'll approve

**Comment by adityapurohit1771** [comment]
> Good attempt. Make some of the changes i've requested and i'll approve

Updated with required Changes @kushal-ti

**Comment by kushal-ti** [comment]
looks good

**Pull request closed without merging**